### PR TITLE
2675 using codeengine with paketo buildpack for liberty

### PR DESCRIPTION
**Pull request by kevin-ortega**
https://blogs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/blog/2022/11/11/using-codeengine-with-paketo-buildpack-for-liberty.html

### CODE REVIEW

**Comment by kevin-ortega** [comment]
#2736 

 @GraceJansen Please review
@dmuelle Can you please review?

**Comment by kevin-ortega** [comment]
https://blogs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/blog/2022/11/11/using-codeengine-with-paketo-buildpack-for-liberty.html

**Comment by kevin-ortega** [comment]
image
image
image
image

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+open-graph-image: https://openliberty.io/img/twitter_card.jpg
+open-graph-image-alt: Open Liberty Logo
+---
+= Deploy Java applications to IBM Cloud Code Engine using the Paketo Buildpack for Liberty
+Kevin Ortega <https://github.com/kevin-ortega>
+:imagesdir: /
+:url-prefix:
+:url-about: /
+
+link:https://www.ibm.com/cloud/code-engine[IBM Cloud Code Engine] provides the next level of easy application deployment and mature management of critical business applications in IBM Cloud. Use Code Engine to easily deploy Java applications using the link:https://openliberty.io/blog/2022/04/01/cloud-native-liberty-buildpack.html[Paketo Buildpack for Liberty] with your container images or source code. With the Paketo Buildpack for Liberty you can use Open Liberty or WebSphere Liberty to run your application.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
You can use link:https://www.ibm.com/cloud/code-engine[IBM Cloud Code Engine] to easily deploy Java applications by using the link:https://openliberty.io/blog/2022/04/01/cloud-native-liberty-buildpack.html[Paketo Buildpack for Liberty] with your container images or source code. Code Engine provides the next level of easy application deployment and mature management of critical business applications in IBM Cloud. 
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+---
+= Deploy Java applications to IBM Cloud Code Engine using the Paketo Buildpack for Liberty
+Kevin Ortega <https://github.com/kevin-ortega>
+:imagesdir: /
+:url-prefix:
+:url-about: /
+
+link:https://www.ibm.com/cloud/code-engine[IBM Cloud Code Engine] provides the next level of easy application deployment and mature management of critical business applications in IBM Cloud. Use Code Engine to easily deploy Java applications using the link:https://openliberty.io/blog/2022/04/01/cloud-native-liberty-buildpack.html[Paketo Buildpack for Liberty] with your container images or source code. With the Paketo Buildpack for Liberty you can use Open Liberty or WebSphere Liberty to run your application.  
+
+The intent of this blog is not a Code Engine tutorial but to provide instructions on how to use the Paketo Buildpack for Liberty with Code Engine.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
This post provides instructions on how to use the Paketo Buildpack for Liberty with Code Engine. To learn more about using Code Engine, see link:https://cloud.ibm.com/docs/codeengine?topic=codeengine-getting-started[Getting started with IBM Cloud Code Engine].

With the Paketo Buildpack for Liberty, you can use Open Liberty or WebSphere Liberty to run your application. 
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+link:https://www.ibm.com/cloud/code-engine[IBM Cloud Code Engine] provides the next level of easy application deployment and mature management of critical business applications in IBM Cloud. Use Code Engine to easily deploy Java applications using the link:https://openliberty.io/blog/2022/04/01/cloud-native-liberty-buildpack.html[Paketo Buildpack for Liberty] with your container images or source code. With the Paketo Buildpack for Liberty you can use Open Liberty or WebSphere Liberty to run your application.  
+
+The intent of this blog is not a Code Engine tutorial but to provide instructions on how to use the Paketo Buildpack for Liberty with Code Engine. 
+
+== What do you need?
+* An https://www.ibm.com/cloud[IBM Cloud] Account
+* Your application source
+
+To deploy container images you'll also need:
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
To deploy container images, you'll also need:
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+* The link:https://buildpacks.io/docs/tools/pack/[pack CLI]
+* https://www.docker.com/[Docker]
+* An externally accessible container registry that can host your images. 
+
+Details on deploying your application from repository source code from the Code Engine console can be found link:https://cloud.ibm.com/docs/codeengine?topic=codeengine-app-source-code#deploy-app-source-code[here].
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
For details on deploying your application from repository source code from the Code Engine console, see  link:https://cloud.ibm.com/docs/codeengine?topic=codeengine-app-source-code#deploy-app-source-code[Deploying your app from repository source code from the console].
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+Details on deploying your application from repository source code with the CLI can be found link:https://cloud.ibm.com/docs/codeengine?topic=codeengine-app-source-code#deploy-app-source-code-cli[here].
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
For details on deploying your application from repository source code with the CLI, see  link:https://cloud.ibm.com/docs/codeengine?topic=codeengine-app-source-code#deploy-app-source-code-cli[Deploying your app from repository source code with the CLI].
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+== Code Engine build strategies
+Code Engine can build your container from source with link:https://paketo.io[Paketo buildpacks] using the link:https://cloud.ibm.com/docs/codeengine?topic=codeengine-plan-build#build-buildpack-strat[Cloud Native Buildpacks] build strategy. The Paketo Buildpack for Liberty is included in the composite Paketo Buildpack for Java that also includes `tomcat` and `tomee` application servers. Tomcat is the default application server and to use Open Liberty requires a little extra configuration. The application server that the Paketo Buildpack for Java will use is influenced by the environment variable `BP_JAVA_APP_SERVER`.  If undefined, the order of Java application servers in the buildpack dictates which one will be picked. Normally, environment variables are provided to buildpacks using the `--env` parameter on the `pack build` CLI command. Code Engine does not use `pack` so the environment variables have to be provided in a `.buildenv` file in the application's root source directory.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
Code Engine can build your container from source with link:https://paketo.io[Paketo buildpacks] using the link:https://cloud.ibm.com/docs/codeengine?topic=codeengine-plan-build#build-buildpack-strat[Cloud Native Buildpacks] build strategy. The Paketo Buildpack for Liberty is included in the composite Paketo Buildpack for Java that also includes `tomcat` and `tomee` application servers. Tomcat is the default application server and to use Open Liberty requires a little extra configuration. 

You can set the application server that the Paketo Buildpack for Java uses by specifying the `BP_JAVA_APP_SERVER` environment variable. If this variable is undefined, the order of Java application servers in the buildpack dictates which one will be picked. Normally, environment variables are provided to buildpacks by using the `--env` parameter on the `pack build` CLI command. However, Code Engine does not use `pack,` so the environment variables must be provided in a `.buildenv` file in the application's root source directory.  
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+== Let's get started
+
+Create the `.buildenv` file in the application's root source directory. At a minimum, `.buildenv` should contain `BP_JAVA_APP_SERVER=liberty` but can contain other environment variables consumed by buildpacks.
+```
+BP_JAVA_APP_SERVER=liberty
+```
+
+Login to the IBM Cloud console and head to the link:https://cloud.ibm.com/codeengine/overview[Code Engine overview] page.  When building from source, enter the url of the respository hosting your source and press `Start creating  ->`.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
Log in to the IBM Cloud console and head to the link:https://cloud.ibm.com/codeengine/overview[Code Engine overview] page.  When you are building from source, enter the URL of the repository that hosts your source and click **Start creating **.
```

**Review by hibell — DISMISSED** [review]
LGTM

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+[.img_border_light]
+image:/img/blog/code_engine_overview.png[IBM Cloud Code Engine Overview,align="center"]
+
+On the `Start creating` page it's important to enter `9080` for `Listening port`.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
On the **Start creating** page, you must enter `9080` as the value for `Listening port`.
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+[.img_border_light]
+image:img/blog/code_engine_start_creating.png[Start creating,width=70%,align="center"] 
+
+We recommend setting the minimum number of instances to `1` in the `Runtime settings` section.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
In the **Runtime settings** section, we recommend setting the minimum number of instances to `1`.
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+[.img_border_light]
+image:img/blog/code_engine_min_instances.png[Minimum instances,width=70%,align="center"] 
+
+Press `Specify build details`.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
Click **Specify build details**. 
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+On the `Start creating` page it's important to enter `9080` for `Listening port`.
+[.img_border_light]
+image:img/blog/code_engine_start_creating.png[Start creating,width=70%,align="center"] 
+
+We recommend setting the minimum number of instances to `1` in the `Runtime settings` section.
+[.img_border_light]
+image:img/blog/code_engine_min_instances.png[Minimum instances,width=70%,align="center"] 
+
+Press `Specify build details`. 
+In this example, the source root is the `finish` directory and contains the `.buildenv` file.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
In this example, the source root that contains the`.buildenv` file is the `finish` directory. 
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc (hunk 8 above)
Comment (dmuelle) [reply posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]: Do we intedn for them to use the sample image? If so, might be helpful to say

"Use the sample image"

or similar so users arent left to wonder what to do with the field

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+image:img/blog/code_engine_min_instances.png[Minimum instances,width=70%,align="center"] 
+
+Press `Specify build details`. 
+In this example, the source root is the `finish` directory and contains the `.buildenv` file. 
+[.img_border_light]
+image:/img/blog/code_engine_build_details.png[Build details, align="center]
+[.img_border_light]
+image:/img/blog/code_engine_git_repo.png[Root of the application's source directory,width=70%,align="center]
+
+Press `Next` and select the `Cloud native buildpack` strategy.  Press `Next`.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
Click **Next** and select **Cloud Native Buildpack** as the source.  Click **Next**.
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+[.img_border_light]
+image:/img/blog/code_engine_build_strategy.png[Build strategy,align="center"]
+
+Enter the details of where Code Engine will store your image and press `Done`.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
Enter the details of where Code Engine will store your image and click **Done**.
```

Comment (dmuelle) [reply posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]: will the user know what to enter here?

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+Press `Create` to have Code Engine create and deploy your application.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
Click **Create** to have Code Engine create and deploy your application.  
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+drwxr-xr-x  12 kevin  staff   384 Oct 19 11:07 ..
+-rw-r--r--@  1 kevin  staff    26 Nov  4 12:53 .buildenv
+-rw-r--r--   1 kevin  staff   790 Oct 19 11:07 Dockerfile
+-rw-r--r--@  1 kevin  staff  4066 Oct 19 11:07 pom.xml
+drwxr-xr-x   4 kevin  staff   128 Oct 19 11:07 src
+drwxr-xr-x  11 kevin  staff   352 Oct 20 08:53 target
+```
+
+.To summarize, two important configuration steps are required to use the Paketo Buildpack for Liberty in Code Engine:
+. Create a `.buildenv` file with `BP_JAVA_APP_SERVER=liberty`
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
. Create a `.buildenv` file that defines the  `BP_JAVA_APP_SERVER=liberty` environment variable.
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+. Set the listening port to 9080
+
+= Other configurations using the `.buildenv` file
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
== Other environment variables you can set in the `.buildenv` file
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+`BP_LIBERTY_INSTALL_TYPE` link:https://github.com/paketo-buildpacks/liberty#install-types[Install type] of Liberty, Open Liberty (ol) or WebSphere Liberty (wlp).  Open Liberty is the default.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
BP_LIBERTY_INSTALL_TYPE::
Specifies the  link:https://github.com/paketo-buildpacks/liberty#install-types[Install type] of Liberty. Possible values are `ol` (Open Liberty) or  `wlp` (WebSphere Liberty).  Open Liberty is the default.
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+`BP_LIBERTY_PROFILE` to specify which liberty profile to install. Valid profiles for Open and WebSphere Liberty are documented link:https://github.com/paketo-buildpacks/liberty#profiles[here].
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
BP_LIBERTY_PROFILE::
Specifies which liberty profile to install. Valid profiles for Open and WebSphere Liberty are documented link:https://github.com/paketo-buildpacks/liberty#profiles[in the buildpacks documentation].
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+
+`BP_LIBERTY_FEATURES` a space separated list of Liberty features to be installed with the Liberty runtime. Supports any valid Liberty feature.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
BP_LIBERTY_FEATURES::
Specifies a space-separated list of Liberty features to be installed with the Liberty runtime. Supports any valid Liberty feature.
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+
+== Taking full advantage of all what the Paketo Buildpack for Liberty has to offer
+Code Engine doesn't use the `pack build` CLI to create container images. It plugs into the cloud native buildpack's lifecycle natively. As a result, some features of the Paketo Buildpack for Liberty are not easily available to Code Engine. Features like link:https://github.com/paketo-buildpacks/liberty/blob/main/docs/installing-ifixes.md[installing iFixes] and link:https://github.com/paketo-buildpacks/liberty#using-custom-features[custom features], and installing from a link:https://github.com/paketo-buildpacks/liberty#building-from-a-packaged-server[packaged server] or link:https://github.com/paketo-buildpacks/liberty#building-from-a-liberty-server[server directory] can't be utilized using Code Engine to create the container image. For these features, you can use the `pack build` CLI to create the container image, push the image to an external container registry and use Code Engine to deploy and manage your container by pulling your container image from the container registry from the Code Engine console or CLI.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
Code Engine doesn't use the `pack build` CLI to create container images. It plugs into the cloud native buildpack's lifecycle natively. As a result, some features of the Paketo Buildpack for Liberty are not easily available to Code Engine. Features like link:https://github.com/paketo-buildpacks/liberty/blob/main/docs/installing-ifixes.md[installing iFixes], link:https://github.com/paketo-buildpacks/liberty#using-custom-features[custom features], and installing from a link:https://github.com/paketo-buildpacks/liberty#building-from-a-packaged-server[packaged server] or link:https://github.com/paketo-buildpacks/liberty#building-from-a-liberty-server[server directory] aren't available when you use Code Engine to create the container image. 

For these features, you can use the `pack build` CLI to create the container image, push the image to an external container registry. Then, use Code Engine to deploy and manage your container by pulling your container image from the container registry from the Code Engine console or CLI.  
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+link:https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-crimage[Follow these instructions] to deploy applications from the IBM Cloud Container Registry.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
link:https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-crimage[Follow these instructions] to deploy applications from the IBM Cloud Container Registry with Code Engine. 
```

**Review by dmuelle — CHANGES_REQUESTED** [review]
ID Review- left some suggestions. 

Generally speaking, it could be helpful to briefly state in the introduction some of the benefits to using this approach. I think the post does a good job of telling you how to do it, but it doesn't say much about why you might want to or in what context it's a helpful strategy.

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+layout: post
+title: "Deploy Java applications to IBM Cloud Code Engine using the Paketo Buildpack for Liberty"
+# Do NOT change the categories section
+categories: blog
+author_picture: https://avatars3.githubusercontent.com/kevin-ortega
+author_github: https://github.com/kevin-ortega
+seo-title:  Deploy Java applications to IBM Cloud Code Engine (ICCE) using the Paketo Buildpack for Liberty - OpenLiberty.io
+seo-description: Learn to deploy your applications to ICCE easily with the Paketo Buildpack for Liberty
+blog_description: "Learn to deploy your applications to ICCE easily with the Paketo Buildpack for Liberty"
+open-graph-image: https://openliberty.io/img/twitter_card.jpg
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
open-graph-image:https://openliberty.io/img/blog/code_engine_build_step_details.png
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+open-graph-image-alt: Open Liberty Logo
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
open-graph-image-alt: screen capture of Code Engine build details
```

File: posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+On the **Start creating** page, you must enter `9080` as the value for `Listening port`.
+[.img_border_light]
+image:img/blog/code_engine_start_creating.png[Start creating,width=70%,align="center"] 
+
+In the **Runtime settings** section, we recommend setting the minimum number of instances to `1`.
+[.img_border_light]
+image:img/blog/code_engine_min_instances.png[Minimum instances,width=70%,align="center"] 
+
+Click **Specify build details**. 
+In this example, the source root that contains the`.buildenv` file is the `finish` directory.
```
Comment (dmuelle) [inline posts/2022-11-11-using-codeengine-with-paketo-buildpack-for-liberty.adoc]:
```suggestion
In this example, the source root that contains the `.buildenv` file is the `finish` directory. 
```

**Review by dmuelle — APPROVED** [review]

**Pull request merged**